Support wildcards when ignoring baseline TFMs

ViktorHofer opened the pull request:
The `PackageValidationBaselineFrameworkToIgnore` property now supports the glob '*' character at the end of the string.

Added tests to verify the behavior. The comparison is still performed with ignored casing and ignored culture.

ViktorHofer committed:
Support glob patterns when ignoring baseline TFMs

Change the task and targets input to a property instead of an item.
This feature hasn't yet shipped nor got documented so it doesn't require a
breaking change noticed.

The new property ' PackageValidationBaselineFrameworksToIgnore' now
supports the glob '*' character. Multiple entries should be separated
via the ';' character. This is documented in both frontends.

Added tests to verify the behavior. The comparison is still performed
with ignored casing and ignored culture.

ericstj commented:
Adding wild-card support here makes me nervous.  We don't want folks to accidentally disable testing without realizing it.  Can you show the sort of usage you're trying to allow?

ericstj commented on src/Compatibility/ApiCompat/Microsoft.DotNet.PackageValidation/Filtering/TargetFrameworkRegexFilter.cs:
```diff
@@ -0,0 +1,65 @@
+#if NET
+                | RegexOptions.NonBacktracking
+#endif
+                );
```
Include a timeout here.

ericstj commented on src/Compatibility/ApiCompat/Microsoft.DotNet.PackageValidation/Filtering/TargetFrameworkRegexFilter.cs:
```diff
@@ -0,0 +1,65 @@
+#else
+            string pattern = patterns.Split(new char[] { TargetFrameworkDelimiter }, StringSplitOptions.RemoveEmptyEntries)
+#endif
+                .Select(p => Regex.Escape(p).Replace("\\*", ".*"))
```
Be sure to include a comment that describes that you're escaping all other regex syntax *except* for `*` which is transformed to `.*`.

ericstj commented on src/Compatibility/ApiCompat/Microsoft.DotNet.PackageValidation/Filtering/TargetFrameworkRegexFilter.cs:
```diff
@@ -0,0 +1,65 @@
+        public IReadOnlyCollection<string> FoundExcludedTargetFrameworks => _foundExcludedTargetFrameworks;
+
+        /// <inheritdoc/>
+        public bool IsExcluded(string targetFramework)
```
Would it make sense to have a `NuGetFramework` overload as well to ensure you're always comparing the same string format rather than relying on the caller to convert to string in the same way?

ViktorHofer replied:
Great idea. Done.

ericstj commented on src/Compatibility/ApiCompat/Microsoft.DotNet.PackageValidation/Validators/BaselinePackageValidator.cs:
```diff
@@ -117,6 +117,15 @@ public void Validate(PackageValidatorOption options)
+            if (options.BaselinePackageFrameworkFilter is not null &&
+                options.BaselinePackageFrameworkFilter.FoundExcludedTargetFrameworks.Count > 0)
+            {
+                log.LogMessage(ApiSymbolExtensions.Logging.MessageImportance.Normal,
```
Consider a higher level of importance if we aren't emitting any suppressions.  Try to increase the visibility of the omitted testing.

Is there anything more we might want to do to catch a case where someone might accidentally omit comparisons?

ViktorHofer replied:
`Normal` message importance is the highest we can go here. We don't ever want to emit messages with `High` importance outside of a failure path.

ericstj replied:
I see.  If all frameworks are excluded should we emit a warning?  If frameworks are excluded from the baseline that are present explicitly in the new package should we emit a warning?  Just brainstorming some potential pitfalls here and how we might mitigate them.  If you think these are good suggestions I'd be OK with us handling those or others separately.

ViktorHofer replied:
I thought I replied to that yesterday. Yes that's a great idea.

ViktorHofer committed:
Use string matching instead of Regex

ViktorHofer committed:
Rename types

ViktorHofer commented:
I changed the implementation to not use Regex underneath and just support the wildcard character at the end of the string. In addition, I reverted the frontend changes that changed to a property instead of items. With that, repositories don't need to react to this change anymore.

ViktorHofer committed:
Add NuGetFramework overload

ericstj commented on src/Compatibility/ApiCompat/Microsoft.DotNet.PackageValidation/Filtering/TargetFrameworkFilter.cs:
```diff
@@ -0,0 +1,61 @@
+            foreach (string excludedTargetFramework in excludedTargetFrameworks)
+            {
+                // Wildcard match
+                if (excludedTargetFramework.Length > 1 && excludedTargetFramework.EndsWith("*"))
```
nit:
```suggestion
                if (excludedTargetFramework.Length > 1 && excludedTargetFramework.EndsWith('*'))
```

ViktorHofer replied:
That overload isn't available on .NET Framework: https://apisof.net/catalog/fa804050-085a-5b5e-3d46-f9fdefd367c6

I will change this to just look at the last character.

ericstj commented on src/Compatibility/ApiCompat/Microsoft.DotNet.PackageValidation/Filtering/TargetFrameworkFilter.cs:
```diff
@@ -0,0 +1,61 @@
+                // Wildcard match
+                if (excludedTargetFramework.Length > 1 && excludedTargetFramework.EndsWith("*"))
+                {
+                    string excludedTargetFrameworkWithoutWildcard = excludedTargetFramework.Substring(0, excludedTargetFramework.Length - 1);
```
if someone passes in just `*` it looks like this would do the right thing (empty string - then startswith(empty)), but we should consider if want to allow that or not and add a test case either way.

carlossanlop replied:
Are there tests for this?

ViktorHofer replied:
I intentionally didn't want to allow that and wanted to guard the substring call which is why I added the `Length > 1` condition. I will add a test case to protect that decision.

ericstj reviewed (approved):
Looks reasonable.  Made a couple suggestions.

carlossanlop commented on test/Microsoft.DotNet.PackageValidation.Tests/Filtering/TargetFrameworkFilterTests.cs:
```diff
@@ -0,0 +1,87 @@
+
+namespace Microsoft.DotNet.PackageValidation.Tests.Filtering
+{
+    public class TargetFrameworkFilterTests
```
Please add a test for passing "*" as suggested by Eric: https://github.com/dotnet/sdk/pull/38374/files#r1473573127

carlossanlop commented on test/Microsoft.DotNet.PackageValidation.Tests/Filtering/TargetFrameworkFilterTests.cs:
```diff
@@ -0,0 +1,87 @@
+namespace Microsoft.DotNet.PackageValidation.Tests.Filtering
+{
+    public class TargetFrameworkFilterTests
+    {
```
Can you also please add a test for the exact case that prompted the creation of this PR? `portable-net45+win8+wpa81+wp8`

Relevant thread: https://github.com/dotnet/maintenance-packages/pull/38/files#r1471429706

carlossanlop replied:
An additional test in the BaselinePackageValidatorTests.cs file would make sense too for this case, since that's the specific scenario that prompted this PR.

ViktorHofer committed:
Add tests for portable and nit feedback on exclusion logic

ViktorHofer committed:
Add logging for excluded baseline targetframework that exists in the current package
